In [15]:
import pandas as pd

visits = pd.read_csv('visits.csv', parse_dates=[1])
cart = pd.read_csv('cart.csv', parse_dates=[1])
checkout = pd.read_csv('checkout.csv', parse_dates=[1])
purchase = pd.read_csv('purchase.csv', parse_dates=[1])

## Funnel for Cool T-Shirts Inc.

In [16]:
# inspectimg the dataframes
print(visits.head())
print(cart.head())
print(checkout.head())
print(purchase.head())

                                user_id          visit_time
0  943647ef-3682-4750-a2e1-918ba6f16188 2017-04-07 15:14:00
1  0c3a3dd0-fb64-4eac-bf84-ba069ce409f2 2017-01-26 14:24:00
2  6e0b2d60-4027-4d9a-babd-0e7d40859fb1 2017-08-20 08:23:00
3  6879527e-c5a6-4d14-b2da-50b85212b0ab 2017-11-04 18:15:00
4  a84327ff-5daa-4ba1-b789-d5b4caf81e96 2017-02-27 11:25:00
                                user_id           cart_time
0  2be90e7c-9cca-44e0-bcc5-124b945ff168 2017-11-07 20:45:00
1  4397f73f-1da3-4ab3-91af-762792e25973 2017-05-27 01:35:00
2  a9db3d4b-0a0a-4398-a55a-ebb2c7adf663 2017-03-04 10:38:00
3  b594862a-36c5-47d5-b818-6e9512b939b3 2017-09-27 08:22:00
4  a68a16e2-94f0-4ce8-8ce3-784af0bbb974 2017-07-26 15:48:00
                                user_id       checkout_time
0  d33bdc47-4afa-45bc-b4e4-dbe948e34c0d 2017-06-25 09:29:00
1  4ac186f0-9954-4fea-8a27-c081e428e34e 2017-04-07 20:11:00
2  3c9c78a7-124a-4b77-8d2e-e1926e011e7d 2017-07-13 11:38:00
3  89fe330a-8966-4756-8f7c-3bdbcd47279a 

In [17]:
# Combining visits and cart using a left merge.
visits_cart = pd.merge(visits, cart, how='left')

print(len(visits_cart))

#  Timestamps that are null for the column cart_time
null_cart = visits_cart[visits_cart.cart_time.isnull()].user_id.count()
print(null_cart) # these are the number of users who visited Cool T-Shirts Inc. but ended up not placing a t-shirt in their cart

# Percentage of users who visited Cool T-Shirts Inc. but ended up not placing a t-shirt in their cart
percent_null_cart = null_cart / float(len(visits_cart))
print(percent_null_cart)

2000
1652
0.826


In [18]:
cart_checkout = pd.merge(cart, checkout, how='left')
print(len(cart_checkout))

null_checkout = len(cart_checkout[cart_checkout.checkout_time.isnull()])
print(null_checkout)

#percentage of users that put items in their cart but didn't checkout
percent_null_ckeckout = null_checkout / float(len(cart_checkout))
print(percent_null_ckeckout)


482
122
0.25311203319502074


In [19]:
#7 Merging all the dataframes in order using a series of left merges and saving it to the variable all_data
all_data = visits.merge(cart, how = 'left').merge(checkout, how='left').merge(purchase, how = 'left')
print(all_data.head())

                                user_id          visit_time  \
0  943647ef-3682-4750-a2e1-918ba6f16188 2017-04-07 15:14:00   
1  0c3a3dd0-fb64-4eac-bf84-ba069ce409f2 2017-01-26 14:24:00   
2  6e0b2d60-4027-4d9a-babd-0e7d40859fb1 2017-08-20 08:23:00   
3  6879527e-c5a6-4d14-b2da-50b85212b0ab 2017-11-04 18:15:00   
4  a84327ff-5daa-4ba1-b789-d5b4caf81e96 2017-02-27 11:25:00   

            cart_time       checkout_time       purchase_time  
0                 NaT                 NaT                 NaT  
1 2017-01-26 14:44:00 2017-01-26 14:54:00 2017-01-26 15:08:00  
2 2017-08-20 08:31:00                 NaT                 NaT  
3                 NaT                 NaT                 NaT  
4                 NaT                 NaT                 NaT  


In [20]:
checkout_purchase = pd.merge(checkout, purchase, how='left')

null_purchase = len(checkout_purchase[checkout_purchase.purchase_time.isnull()])

# Percentage of users proceeded to checkout, but did not purchase a t-shirt
percent_null_purchase = null_purchase / float(len(checkout_purchase))
print(percent_null_purchase)

0.1688963210702341


Which step of the funnel is weakest (i.e., has the highest percentage of users not completing it)? <br>
Ans: Percentage of users who visited Cool T-Shirts Inc. but ended up not placing a t-shirt in their cart (82.6%)

## Average Time to Purchase

In [21]:
all_data['time_to_purchase'] = all_data.purchase_time - all_data.visit_time

# print(all_data.time_to_purchase)

print(all_data.time_to_purchase.mean()) # calculating the average time to purchase

0 days 00:43:53.360160965
